core/src/main/java/org/elasticsearch/search/aggregations/AggregatorFactories.java

```diff
@@ -251,4 +259,97 @@ private void resolvePipelineAggregatorOrder(Map<String, AggregatorFactory> aggFa
+        for (int i = 0; i < pipelineFactoriesSize; i++) {
+            PipelineAggregatorFactory factory = in.readPipelineAggregatorFactory();
+            pipelineAggregatorFactoriesList.add(factory);
+        }
+        AggregatorFactories aggregatorFactories = new AggregatorFactories(
+                factoriesList.toArray(new AggregatorFactory[factoriesList.size()]), pipelineAggregatorFactoriesList);
+        boolean hasParent = in.readBoolean();
+        if (hasParent) {
+            AggregatorFactory parent = in.readAggregatorFactory();
+            aggregatorFactories.parent = parent;
+        }
+        return aggregatorFactories;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVInt(this.factories.length);
+        for (AggregatorFactory factory : factories) {
+            out.writeAggregatorFactory(factory);
+        }
+        out.writeVInt(this.pipelineAggregatorFactories.size());
+        for (PipelineAggregatorFactory factory : pipelineAggregatorFactories) {
+            out.writePipelineAggregatorFactory(factory);
+        }
+        boolean hasParent = parent != null;
+        out.writeBoolean(hasParent);
+        if (hasParent) {
+            out.writeAggregatorFactory(parent);
+        }
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        if (factories != null) {
+            for (AggregatorFactory subAgg : factories) {
+                subAgg.toXContent(builder, params);
+            }
+        }
+        if (pipelineAggregatorFactories != null) {
+            for (PipelineAggregatorFactory subAgg : pipelineAggregatorFactories) {
+                subAgg.toXContent(builder, params);
+            }
+        }
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
+        int result = 1;
+        result = prime * result + Arrays.hashCode(factories);
+        result = prime * result + ((parent == null) ? 0 : parent.hashCode());
+        result = prime * result + ((pipelineAggregatorFactories == null) ? 0 : pipelineAggregatorFactories.hashCode());
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj)
+            return true;
+        if (obj == null)
+            return false;
+        if (getClass() != obj.getClass())
+            if (!(obj.getClass().isAssignableFrom(getClass()) || getClass().isAssignableFrom(obj.getClass())))
+                return false;
+        AggregatorFactories other = (AggregatorFactories) obj;
+        if (!Arrays.equals(factories, other.factories))
+            return false;
+        if (parent == null) {
+            if (other.parent != null)
+                return false;
```

Comment:
My theory here was that you could have the same aggs structure (names, sub-aggs, options) in two different parts of the aggregation tree and they wouldn't be equal because they had different parents. So equality basically included the position in the aggs tree. But maybe thats not really right because then we are using an outside property (position in the aggs tree) to determine equality of an object?